ua/schema.go

```diff
@@ -0,0 +1,137 @@
+package ua
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pascaldekloe/name"
+)
+
+// Serializable checks if the Go struct t can be serialized into the UA type name.
+func Serializable(t reflect.Type, id string, types map[string]*StructureDefinition) error {
+	uatyp, err := UAType(id, types)
+	if err != nil {
+		return err
+	}
+	uadef, godef := GoString(uatyp), GoString(anonType(t))
+	if uadef != godef {
+		return fmt.Errorf(`schema: Go type "%s" is not serializable into UA type "%s" (%s)`, godef, uadef, id)
+	}
+	return nil
+}
+
+// GoString returns a Go-syntax representation of the type of the value.
+func GoString(t reflect.Type) string {
+	return fmt.Sprintf("%T", reflect.New(t).Interface())
+}
+
+// anonType returns the struct without the name.
+func anonType(t reflect.Type) reflect.Type {
+	switch {
+	// do not unpack time.Time
+	case isTimeType(t):
+		return t
+	case t.Kind() == reflect.Slice:
+		return reflect.SliceOf(anonType(t.Elem()))
+	case t.Kind() == reflect.Struct:
+		return reflect.StructOf(structFields(t))
+	default:
+		return t
+	}
+}
+
+// structFields returns the fields of a struct as an array.
+// it panics if t is not a struct.
+func structFields(t reflect.Type) []reflect.StructField {
+	if t.Kind() != reflect.Struct {
+		panic(fmt.Sprintf("%s is not a struct", t.Name()))
+	}
+	var fields []reflect.StructField
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		f.Name = name.CamelCase(f.Name, true)
+		f.Type = anonType(f.Type)
+		fields = append(fields, f)
+	}
+	return fields
+}
+
+// uaType returns an anonymous Go representation of the OPC/UA struct definition with the NodeID 'id'.
+func UAType(id string, types map[string]*StructureDefinition) (reflect.Type, error) {
+	t := types[id]
+	if t == nil {
+		return nil, fmt.Errorf("schema: ua type %q not defined", id)
+	}
+
+	// we don't support structs with optional fields or unions (yet or maybe never)
+	if t.StructureType != StructureTypeStructure {
+		return nil, fmt.Errorf("schema: ua type %q has unsupported structure type %s", id, t.StructureType)
+	}
+
+	var fields []reflect.StructField
+	for _, f := range t.Fields {
+		if len(f.ArrayDimensions) > 1 {
+			return nil, fmt.Errorf("schema: ua type %s.%s is a multi-dimensional array: %v", id, f.Name, f.ArrayDimensions)
```

Comment:
This error is unclear. It just states the fact without saying what is wrong imo  :)
I am not sure i understand why is it an error. This is a valid use case for PackML for example.